airflow/models/sensorinstance.py

```diff
@@ -0,0 +1,166 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import json
+
+from sqlalchemy import BigInteger, Column, Index, Integer, String, Text
+
+from airflow.configuration import conf
+from airflow.exceptions import AirflowException
+from airflow.models.base import ID_LEN, Base
+from airflow.utils import timezone
+from airflow.utils.session import provide_session
+from airflow.utils.sqlalchemy import UtcDateTime
+from airflow.utils.state import State
+
+
+class SensorInstance(Base):
+    """
+    SensorInstance support the smart sensor service. It stores the sensor task states
+    and context that required for poking include poke context and execution context.
+    In sensor_instance table we also save the sensor operator classpath so that inside
+    smart sensor there is no need to import the dagbag and create task object for each
+    sensor task.
+
+    SensorInstance include another set of columns to support the smart sensor shard on
+    large number of sensor instance. By hashcode generated from poke_contex and shardcode
+    the distributed smart sensor processes can work on different shards.
+
+    """
+
+    __tablename__ = "sensor_instance"
+
+    id = Column(Integer, primary_key=True)
+    task_id = Column(String(ID_LEN), nullable=False)
+    dag_id = Column(String(ID_LEN), nullable=False)
+    execution_date = Column(UtcDateTime, nullable=False)
+    state = Column(String(20))
+    _try_number = Column('try_number', Integer, default=0)
+    start_date = Column(UtcDateTime)
+    operator = Column(String(1000), nullable=False)
+    op_classpath = Column(String(1000), nullable=False)
+    hashcode = Column(BigInteger, nullable=False)
+    shardcode = Column(Integer, nullable=False)
+    poke_context = Column(Text, nullable=False)
+    execution_context = Column(Text)
+    created_at = Column(UtcDateTime, default=timezone.utcnow(), nullable=False)
+    updated_at = Column(UtcDateTime,
+                        default=timezone.utcnow(),
+                        onupdate=timezone.utcnow(),
+                        nullable=False)
+
+    __table_args__ = (
+        Index('ti_primary_key', dag_id, task_id, execution_date, unique=True),
+
+        Index('si_hashcode', hashcode),
+        Index('si_shardcode', shardcode),
+        Index('si_state_shard', state, shardcode),
+        Index('si_updated_at', updated_at),
+    )
+
+    def __init__(self, ti):
+        self.dag_id = ti.dag_id
+        self.task_id = ti.task_id
+        self.execution_date = ti.execution_date
+
+    @staticmethod
+    def get_classpath(obj):
+        """
+        Get the object dotted class path. Used for getting operator classpath
+        :param obj:
```

Comment:
Needs a blank space between description and params for Sphinx rendering

```suggestion

        :param obj:
```